src/nupic/engine/Input.cpp

```diff
@@ -460,35 +465,27 @@ void Input::initialize() {
         << "was called. Region's dimensions must be specified.";
   }
 
-  if (isSparse_) {
-    NTA_CHECK(isRegionLevel_) << "Sparse data must be region level";
-    NTA_CHECK(data_.getType() == NTA_BasicType_UInt32)
-        << "Sparse data must be uint32";
-  }
-
   // Calculate our size and the offset of each link
+  // The offset is location within the input buffer where
+  // this link will place its data in a Fan-In type situation.
+  // The final count will be the entire buffer size.
+  // If there is more than one link to this same input, its a FanIn.
   size_t count = 0;
+  bool is_FanIn = links_.size() > 1;
   for (std::vector<std::shared_ptr<Link>>::const_iterator l = links_.begin();
        l != links_.end(); l++) {
     linkOffsets_.push_back(count);
     // Setting the destination offset makes the link usable.
-    // TODO: change
-    (*l)->initialize(count);
+    (*l)->initialize(count, is_FanIn);
     count += (*l)->getSrc().getData().getCount();
   }
 
   // Later we may optimize with the zeroCopyEnabled_ flag but
-  // for now we always allocate our own buffer.
+  // for now we always allocate our own input buffer.
+  // Create the Input buffer.
```

Comment:
Questions, out of curiosity:  I see some changes to fan-in/fan-out, does it copy or work inplace?

Reply:
>  I see some changes to fan-in/fan-out, does it copy or work inplace?

For fan-in there must be a copy because each output is being written to a different offset in the same input buffer.  Otherwise, all connections are non-copy unless the data type is different OR the link uses a splitter map to remap the bits in the buffer.  Originally all links required a copy.